src/rgw/rgw_rest_pubsub.cc

```diff
@@ -633,33 +820,51 @@ void RGWPSCreateNotif_ObjStore_S3::execute(optional_yield y) {
       return;
     }
     ldpp_dout(this, 20) << "successfully auto-generated notification for unique topic '" << unique_topic_name << "'" << dendl;
-  
-    if (!push_only) {
-      // generate the subscription with destination information from the original topic
-      rgw_pubsub_sub_dest dest = topic_info.dest;
-      dest.bucket_name = data_bucket_prefix + s->owner.get_id().to_str() + "-" + unique_topic_name;
-      dest.oid_prefix = data_oid_prefix + notif_name + "/";
-      auto sub = ps->get_sub(notif_name);
-      op_ret = sub->subscribe(this, unique_topic_name, dest, y, notif_name);
-      if (op_ret < 0) {
-        ldpp_dout(this, 1) << "failed to auto-generate subscription '" << notif_name << "', ret=" << op_ret << dendl;
-        // rollback generated notification (ignore return value)
-        b->remove_notification(this, unique_topic_name, y);
-        // rollback generated topic (ignore return value)
-        ps->remove_topic(this, unique_topic_name, y);
-        return;
-      }
-      ldpp_dout(this, 20) << "successfully auto-generated subscription '" << notif_name << "'" << dendl;
-    }
   }
 }
 
+int RGWPSCreateNotifOp::verify_permission(optional_yield y) {
+  int ret = get_params();
+  if (ret < 0) {
+    return ret;
+  }
+
+  std::unique_ptr<rgw::sal::User> user = store->get_user(s->owner.get_id());
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  ret = store->get_bucket(this, user.get(), s->owner.get_id().tenant, bucket_name, &bucket, y);
+  if (ret < 0) {
+    ldpp_dout(this, 1) << "failed to get bucket info, cannot verify ownership" << dendl;
+    return ret;
+  }
+  bucket_info = bucket->get_info();
+
+  if (bucket_info.owner != s->owner.get_id()) {
+    ldpp_dout(this, 1) << "user doesn't own bucket, not allowed to create notification" << dendl;
+    return -EPERM;
+  }
+  return 0;
+}
+
 // command (extension to S3): DELETE /bucket?notification[=<notification-id>]
-class RGWPSDeleteNotif_ObjStore_S3 : public RGWPSDeleteNotifOp {
-private:
+class RGWPSDeleteNotifOp : public RGWDefaultResponseOp {
+  private:
```

Comment:
no need for 'private' here

Reply:
i know it is the default. by, personally, i prefer this to be more verbose